### PR TITLE
Be careful about where we change braces in closures

### LINKED ISSUE #863: Rustfmt breaks some closures
See [here](http://is.gd/4MRqu3) for an example.

Running rustfmt changes the closure `closure` to match the one marked `broken_closure`, which doesn't compile and breaks builds. Possible solutions I see are not changing the original, or adding parentheses as in `fixed_broken_closure`.

### PR DESCRIPTION
**Pull request by nrc**
And some general refactoring of closure code.

Fixes #863 

r? @marcusklaas

### CODE REVIEW

File: tests/target/chains-no-overflow.rs [hunk 1]
```diff
@@ -9,20 +9,16 @@ fn main() {
                        .ddddddddddddddddddddddddddd
                        .eeeeeeee();
 
-    x().y(|| {
-           match cond() {
-               true => (),
-               false => (),
-           }
+    x().y(|| match cond() {
```
Comment (marcusklaas) [inline tests/target/chains-no-overflow.rs]: I'm not sure this is an improvement readability wise. Could we go with the "fixed_broken_closure" solution of #863?

Comment (nrc) [inline tests/target/chains-no-overflow.rs]: I'm not keen on introducing unnecessary parens - the compiler even lints against it in `if` expressions, etc. The only reason to have the braces is to make a multi-line expression more coherent, and I don't think parens accomplish that as well.

I'm not super-keen on this change either, but I have grown more used to it and I like it more than any of the other alternatives.

File: tests/target/closure.rs [hunk 2]
```diff
@@ -62,9 +60,7 @@ fn main() {
 
     let closure_with_return_type = |aaaaaaaaaaaaaaaaaaaaaaarg1,
                                     aaaaaaaaaaaaaaaaaaaaaaarg2|
-                                    -> Strong {
-        "sup".to_owned()
-    };
+                                    -> Strong { "sup".to_owned() };
```
Comment (marcusklaas) [inline tests/target/closure.rs]: Was this intended?

Comment (nrc) [inline tests/target/closure.rs]: hmmm, kinda, but it probably shouldn't...

**Comment by nrc** [comment]
addressed the dodgy change

**Comment by marcusklaas** [comment]
I guess I too have become a little accustomed to the new style in the past 15 minutes :-P

r+ when travis gives the green light

**Pull request merged**